src/metaswitch/clearwater/queue_manager/etcd_synchronizer.py

```diff
@@ -0,0 +1,253 @@
+            else: # pragma: no cover
+                self._client.write(self.key(), queue_config, prevExist=False)
+        except (EtcdAlreadyExist, ValueError): # pragma: no cover
+            _log.debug("Contention on etcd write")
+            # Our etcd write failed because someone got there before us. We
+            # don't need to retry in this case as we'll just pick up the
+            # changes in the next etcd read
+        except Exception as e:
+            # Catch-all error handler (for invalid requests, timeouts, etc) -
+            # unset our state and start over.
+            _log.error("{} caught {!r} when trying to write {} with index {}"
+                       " - pause before retrying"
+                       .format(self._ip, e, queue_config, self._index))
+
+            # Setting last_cluster_view to None means that the next successful
+            # read from etcd will trigger the state machine, which will mean
+            # that any necessary work/state changes get retried.
+            self._last_value, self._last_index = None, None
+
+            # Sleep briefly to avoid hammering a failed server
+            self.pause()
+            needs_no_retry = False
+     
+        return needs_no_retry
+
+    def remove_from_queue(self, successful):
+        # Get and parse the current value
+        self._last_value, self._index = self.read_from_etcd(wait=False)
+        queue_config = json.loads(self._last_value)
+
+        # What is the current status?
+        queue_status = status_in_queue(queue_config[constants.JSON_QUEUED], self._id)
+        restart_status = status_in_queue(queue_config[constants.JSON_COMPLETED], self._id)
+        error_status = status_in_queue(queue_config[constants.JSON_ERRORED], self._id)
+
+        if len(queue_config[constants.JSON_QUEUED]) > 0 and \
+           queue_config[constants.JSON_QUEUED][0][constants.JSON_ID] == self._id and \
+           queue_config[constants.JSON_QUEUED][0][constants.JSON_STATUS] == constants.S_PROCESSING:
+            del queue_config[constants.JSON_QUEUED][0]
+            
+            if successful:
+                if constants.S_QUEUED not in queue_status and restart_status == []:
+                    add_id_to_json(queue_config[constants.JSON_COMPLETED], self._id, constants.S_DONE)
+            else:
+                if not (len(queue_config[constants.JSON_QUEUED]) > 0 and \
+                        queue_config[constants.JSON_QUEUED][0][constants.JSON_ID] == self._id) and \
+                   constants.S_FAILURE not in error_status:
+                    remove_id_from_json(self._id, queue_config[constants.JSON_ERRORED])
+                    add_id_to_json(queue_config[constants.JSON_ERRORED], self._id, constants.S_FAILURE)
+ 
+            if len(queue_config[constants.JSON_QUEUED]) > 0 and \
+               status_in_queue(queue_config[constants.JSON_ERRORED], queue_config[constants.JSON_QUEUED][0][constants.JSON_ID]) != []:
```

Comment:
It took me a while to figure out that this line is checking whether this ID is in error state - could you make it more readable?